library/ssl_tls.c

```diff
@@ -1990,25 +1958,31 @@ static int ssl_decrypt_buf( mbedtls_ssl_context *ssl )
     {
         unsigned char mac_expect[MBEDTLS_SSL_MAC_ADD];
 
-        ssl->in_msglen -= ssl->transform_in->maclen;
+        if( rec->data_len < transform->maclen )
+        {
+            MBEDTLS_SSL_DEBUG_MSG( 1, ( "msglen (%d) < maclen (%d)",
+                                        rec->data_len, transform->maclen ) );
+            return( MBEDTLS_ERR_SSL_INVALID_MAC );
```

Comment:
Blocker: this is an early return that depends on secret data: `rec->data_len` depends on `padlen` since line 1937.

An option would be to test for this as well near line 1875, using the same condition as before the patch:
```
-        if( ssl->in_msglen < ssl->transform_in->maclen + padlen &&
-            auth_done == 0 )
```
I know it makes things less local but I don't see a better option now. (Setting `correct = 0` here is possible, but we would also need set `padlen = 1` before we subtract it from the data length, or we could try to compensate by adding back something here but that would easily get very messy.)

See also my comment near delete line 1785 (pre-patch number).

Reply:
(Hopefully) fixed.